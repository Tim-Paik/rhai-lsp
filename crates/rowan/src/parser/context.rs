//! The parser context is a separate module to limit
//! the API surface for the parser functions.
#![allow(dead_code)]
use rowan::{GreenNodeBuilder, TextRange, TextSize};

use crate::syntax::{
    Lexer,
    SyntaxKind::{self, *},
};

use super::{Parse, ParseError, ParseErrorKind};

pub(crate) struct Context<'src> {
    lexer: Lexer<'src>,
    current_token: Option<SyntaxKind>,
    last_token: Option<SyntaxKind>,
    green: GreenNodeBuilder<'static>,
    errors: Vec<ParseError>,

    /// Tracks statements being separated by ";".
    statement_closed: bool,
}

impl<'src> Context<'src> {
    pub(crate) fn new(source: &'src str) -> Self {
        Self {
            lexer: Lexer::new(source),
            current_token: None,
            last_token: None,
            green: GreenNodeBuilder::new(),
            errors: Vec::new(),

            statement_closed: true,
        }
    }

    pub(crate) fn token(&mut self) -> Option<SyntaxKind> {
        // Eat insignificant tokens
        loop {
            if self.current_token.is_none() {
                self.current_token = self.lexer.next();
            }

            match self.current_token {
                Some(COMMENT_BLOCK) | Some(COMMENT_LINE) | Some(WHITESPACE) => {
                    self.eat();
                    self.last_token = None // We aren't interested in these.
                }
                Some(ERROR) => {
                    self.eat_error(ParseErrorKind::InvalidInput);
                    self.last_token = None // We aren't interested in these.
                }
                _ => break,
            }
        }

        self.current_token
    }

    pub(crate) fn previous_token(&self) -> Option<SyntaxKind> {
        self.last_token
    }

    pub(crate) fn eat(&mut self) {
        if let Some(t) = self.current_token.take() {
            self.green.token(t.into(), self.lexer.slice());
            self.last_token = Some(t);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_as(&mut self, kind: SyntaxKind) {
        if self.current_token.is_some() {
            self.green.token(kind.into(), self.lexer.slice());
            self.last_token = Some(kind);
        }
        self.current_token = None;
    }

    pub(crate) fn eat_error(&mut self, error: ParseErrorKind) {
        self.add_error(error);
        self.eat();
    }

    pub(crate) fn eat_error_as(&mut self, kind: SyntaxKind, error: ParseErrorKind) {
        self.add_error(error);
        self.eat_as(kind);
    }

    pub(crate) fn add_error(&mut self, error: ParseErrorKind) {
        let span = self.lexer.span();
        self.errors.push(ParseError::new(
            TextRange::new(
                TextSize::from(span.start as u32),
                TextSize::from(span.end as u32),
            ),
            error,
        ));
    }

    pub(crate) fn start_node(&mut self, kind: SyntaxKind) {
        self.green.start_node(kind.into());
    }

    pub(crate) fn finish_node(&mut self) {
        self.green.finish_node();
    }

    pub(crate) fn finish(self) -> Parse {
        Parse {
            errors: self.errors,
            green: self.green.finish(),
        }
    }

    pub(crate) fn statement_closed(&self) -> bool {
        self.statement_closed
    }

    pub(crate) fn set_statement_closed(&mut self, v: bool) {
        self.statement_closed = v;
    }
}
