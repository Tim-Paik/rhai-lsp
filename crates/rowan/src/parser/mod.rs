//! This module contains all parsing-related tools including a
//! recursive-descent parser.

use crate::syntax::{SyntaxKind, SyntaxNode};
use rowan::GreenNode;
use thiserror::Error;

use self::context::Context;

mod context;
pub mod parsers;

/// A flexible parser.
/// 
/// Parsing happens via given parser functions (also found in [`parsers`]).
/// 
/// # Example Uses
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// // Parse some Rhai code (file).
/// let parser = Parser::new(r#"print("hello rhai!")"#);
/// let parse = parser.parse();
/// 
/// assert!(parse.errors.is_empty());
/// ```
/// 
/// ```
/// use rhai_rowan::parser::Parser;
/// use rhai_rowan::parser::parsers::parse_expr;
///
/// // Parse an expression explicitly.
/// let mut parser = Parser::new(r#"print("hello rhai!")"#);
/// parser.execute(parse_expr);
/// let parse = parser.finish();
/// 
/// assert!(parse.errors.is_empty());
/// ```
pub struct Parser<'src> {
    context: Context<'src>,
}

impl<'src> Parser<'src> {
    /// Create a new parser with the given source.
    #[must_use]
    pub fn new(source: &'src str) -> Self {
        Self {
            context: Context::new(source),
        }
    }

    /// Finish parsing.
    /// 
    /// # Panics
    /// 
    /// If no parser function was called,
    /// or a parser was left in an invalid state (bug).
    #[must_use]
    pub fn finish(self) -> Parse {
        self.context.finish()
    }
}

impl<'src> Parser<'src> {
    /// Execute a parser function for the given parser.
    pub fn execute<F: FnOnce(&mut Context)>(&mut self, f: F) {
        f(&mut self.context);
    }
}

/// The result of parsing, containing all errors
/// and the final green tree.
#[derive(Debug, Clone)]
pub struct Parse {
    /// Parse errors.
    pub errors: Vec<ParseError>,
    /// Parsed green tree.
    pub green: GreenNode,
}

impl Parse {
    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn clone_syntax(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }

    /// Turn the result green tree into a CST.
    /// *This ignores errors*, the resulting tree
    /// can be potentially syntactically invalid.
    #[must_use]
    pub fn into_syntax(self) -> SyntaxNode {
        SyntaxNode::new_root(self.green)
    }
}

/// A parse (syntax) error.
#[derive(Debug, Clone, PartialEq, Eq, Error)]
#[error("parse error at {range:?}: {kind}")]
pub struct ParseError {
    /// The span of the error in the parsed source.
    pub range: rowan::TextRange,
    /// Error kind.
    pub kind: ParseErrorKind,
}

impl ParseError {
    fn new(range: rowan::TextRange, kind: ParseErrorKind) -> Self {
        Self { range, kind }
    }
}

/// All the non-fatal parsing errors that can occur.
#[derive(Debug, PartialEq, Eq, Clone, Error)]
pub enum ParseErrorKind {
    #[error("unexpected EOF")]
    UnexpectedEof,

    #[error(r#"invalid input"#)]
    InvalidInput,

    #[error(r#"unexpected token"#)]
    UnexpectedToken,

    #[error(r#"expected token "{0:?}""#)]
    ExpectedToken(SyntaxKind),

    #[error(
        r#"expected one of "{}""#,
        .0
            .iter()
            .map(|s|format!("{:?}", s))
            .collect::<Vec<String>>().join(",")
    )]
    ExpectedOneOfTokens(Vec<SyntaxKind>),
}
