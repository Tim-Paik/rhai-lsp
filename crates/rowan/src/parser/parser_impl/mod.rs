#![deny(unreachable_patterns)]

use super::context::Context;
use crate::parser::ParseErrorKind;
use crate::syntax::SyntaxKind::*;
use crate::T;

macro_rules! require_token {
    ($ctx:ident) => {
        match $ctx.token() {
            Some(t) => t,
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

    ($ctx:ident, finish_node) => {
        match $ctx.token() {
            Some(t) => t,
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::UnexpectedEof);
                return;
            }
        }
    };
}

macro_rules! expect_token {
    ($ctx:ident, finish_node, $($token:tt)*) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => {
                    $ctx.finish_node();
                    $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                    return;
                }
            }
            None => {
                $ctx.finish_node();
                $ctx.eat_error(ParseErrorKind::ExpectedToken($($token)*));
                return;
            }
        }
    };
    ($ctx:ident, $($token:tt)+) => {
        match $ctx.token() {
            Some(t) => match t {
                $($token)* => {
                    $ctx.eat();
                },
                _ => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
            }
            None => return $ctx.eat_error(ParseErrorKind::UnexpectedEof),
        }
    };

}

impl<'src> super::Parser<'src> {
    pub fn parse_file(&mut self) {
        parse_file(&mut self.context);
    }
}

fn parse_file(ctx: &mut Context) {
    ctx.start_node(FILE);
    if let Some(SHEBANG) = ctx.token() {
        parse_shebang(ctx);
    }

    ctx.set_statement_closed(true);
    while ctx.token().is_some() {
        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }
    ctx.set_statement_closed(true);

    ctx.finish_node()
}

fn parse_shebang(ctx: &mut Context) {
    let token = require_token!(ctx);

    if !matches!(token, SHEBANG) {
        return ctx.eat_error(ParseErrorKind::ExpectedToken(SHEBANG));
    }

    ctx.eat()
}

fn parse_stmt(ctx: &mut Context) {
    let token = require_token!(ctx);
    ctx.start_node(STMT);
    ctx.set_statement_closed(false);

    if token == T![";"] {
        ctx.finish_node();
        ctx.set_statement_closed(true);
        return ctx.eat();
    }

    parse_item(ctx);

    if let Some(token) = ctx.token() {
        if token == T![";"] {
            ctx.eat();
            ctx.set_statement_closed(true);
        }
    }

    ctx.finish_node();
}

fn parse_item(ctx: &mut Context) {
    ctx.start_node(ITEM);

    // Parse doc comments if any.
    while matches!(
        require_token!(ctx, finish_node),
        COMMENT_BLOCK_DOC | COMMENT_LINE_DOC
    ) {
        ctx.eat();
    }

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr(ctx: &mut Context) {
    ctx.start_node(EXPR);
    let token = require_token!(ctx, finish_node);

    match token {
        T!["let"] => parse_expr_let(ctx),
        T!["const"] => parse_expr_const(ctx),
        T!["{"] => parse_expr_block(ctx),
        T!["fn"] => parse_expr_fn(ctx),
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => parse_expr_lit(ctx),
        IDENT => parse_expr_ident(ctx),
        _ => ctx.eat_error(ParseErrorKind::UnexpectedToken),
    }

    ctx.finish_node();
}

fn parse_expr_ident(ctx: &mut Context) {
    ctx.start_node(EXPR_IDENT);
    parse_ident(ctx);
    ctx.finish_node();
}

fn parse_expr_lit(ctx: &mut Context) {
    ctx.start_node(EXPR_LIT);
    parse_lit(ctx);
    ctx.finish_node();
}

fn parse_expr_let(ctx: &mut Context) {
    ctx.start_node(EXPR_LET);

    expect_token!(ctx, finish_node, T!["let"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_const(ctx: &mut Context) {
    ctx.start_node(EXPR_CONST);

    expect_token!(ctx, finish_node, T!["const"]);
    expect_token!(ctx, finish_node, T!["ident"]);
    expect_token!(ctx, finish_node, T!["="]);

    parse_expr(ctx);

    ctx.finish_node();
}

fn parse_expr_block(ctx: &mut Context) {
    ctx.start_node(EXPR_BLOCK);

    expect_token!(ctx, finish_node, T!["{"]);

    ctx.set_statement_closed(true);
    loop {
        let token = require_token!(ctx, finish_node);

        if token == T!["}"] {
            break;
        }

        if !ctx.statement_closed() {
            ctx.add_error(ParseErrorKind::ExpectedToken(T![";"]));
        }

        parse_stmt(ctx);
    }

    expect_token!(ctx, finish_node, T!["}"]);

    // Blocks also don't require ";" when used as statements.
    ctx.set_statement_closed(true);

    ctx.finish_node();
}

fn parse_expr_fn(ctx: &mut Context) {
    ctx.start_node(EXPR_FN);

    expect_token!(ctx, finish_node, T!["fn"]);
    expect_token!(ctx, finish_node, T!["ident"]);

    parse_param_list(ctx);
    parse_expr_block(ctx);

    ctx.finish_node();
}

fn parse_param_list(ctx: &mut Context) {
    ctx.start_node(PARAM_LIST);

    let start_token = require_token!(ctx, finish_node);

    match start_token {
        T!["("] | T!["|"] => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![T!["("], T!["|"]]));
            return;
        }
    }

    loop {
        let token = require_token!(ctx, finish_node);
        if matches!(
            (start_token, token),
            (T!["("], T![")"]) | (T!["|"], T!["|"])
        ) {
            ctx.eat();
            break;
        }

        parse_param(ctx);

        let end_token = require_token!(ctx, finish_node);

        match (start_token, end_token) {
            (T!["("], T![")"]) | (T!["|"], T!["|"]) => {
                ctx.eat();
                break;
            }
            (_, T![","]) => {
                ctx.eat();
            }
            _ => {
                ctx.eat_error(ParseErrorKind::ExpectedToken(T![","]));
                break;
            }
        }
    }

    ctx.finish_node();
}

fn parse_param(ctx: &mut Context) {
    ctx.start_node(PARAM);

    expect_token!(ctx, finish_node, T!["ident"]);

    ctx.finish_node();
}

fn parse_lit(ctx: &mut Context) {
    ctx.start_node(LIT);
    let token = require_token!(ctx, finish_node);

    match token {
        LIT_INT | LIT_FLOAT | LIT_BOOL | LIT_STR | LIT_CHAR => {
            ctx.eat();
        }
        _ => {
            ctx.eat_error(ParseErrorKind::ExpectedOneOfTokens(vec![
                LIT_INT, LIT_FLOAT, LIT_BOOL, LIT_STR, LIT_CHAR,
            ]));
        }
    }

    ctx.finish_node();
}

fn parse_ident(ctx: &mut Context) {
    let token = require_token!(ctx);

    match token {
        IDENT => ctx.eat(),
        _ => ctx.eat_error(ParseErrorKind::ExpectedToken(IDENT)),
    }
}

// Expr =
// | ExprUni
// | ExprBin
// | ExprParen
// | ExprArray
// | ExprIndex
// | ExprObject
// | ExprCall
// | ExprMethodCall
// | ExprAccess
// | ExprClosure
// | ExprIf
// | ExprLoop
// | ExprFor
// | ExprWhile
// | ExprBreak
// | ExprContinue
// | ExprSwitch
// | ExprReturn
// | ExprPath
// | ExprImport
