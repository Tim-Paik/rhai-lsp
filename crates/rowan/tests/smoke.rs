use rhai_rowan::parser::{Operator, Parser};
use test_case::test_case;

#[test_case("simple", include_str!("../../../testdata/valid/simple.rhai"))]
#[test_case("array", include_str!("../../../testdata/valid/array.rhai"))]
#[test_case("assignment", include_str!("../../../testdata/valid/assignment.rhai"))]
#[test_case("comments", include_str!("../../../testdata/valid/comments.rhai"))]
#[test_case("fibonacci", include_str!("../../../testdata/valid/fibonacci.rhai"))]
#[test_case("for1", include_str!("../../../testdata/valid/for1.rhai"))]
#[test_case("for2", include_str!("../../../testdata/valid/for2.rhai"))]
#[test_case("function_decl1", include_str!("../../../testdata/valid/function_decl1.rhai"))]
#[test_case("function_decl2", include_str!("../../../testdata/valid/function_decl2.rhai"))]
#[test_case("function_decl3", include_str!("../../../testdata/valid/function_decl3.rhai"))]
#[test_case("function_decl4", include_str!("../../../testdata/valid/function_decl4.rhai"))]
#[test_case("if1", include_str!("../../../testdata/valid/if1.rhai"))]
#[test_case("if2", include_str!("../../../testdata/valid/if2.rhai"))]
#[test_case("loop", include_str!("../../../testdata/valid/loop.rhai"))]
#[test_case("mat_mul", include_str!("../../../testdata/valid/mat_mul.rhai"))]
#[test_case("module", include_str!("../../../testdata/valid/module.rhai"))]
#[test_case("oop", include_str!("../../../testdata/valid/oop.rhai"))]
#[test_case("op1", include_str!("../../../testdata/valid/op1.rhai"))]
#[test_case("op2", include_str!("../../../testdata/valid/op2.rhai"))]
#[test_case("op3", include_str!("../../../testdata/valid/op3.rhai"))]
#[test_case("primes", include_str!("../../../testdata/valid/primes.rhai"))]
#[test_case("speed_test", include_str!("../../../testdata/valid/speed_test.rhai"))]
#[test_case("string", include_str!("../../../testdata/valid/string.rhai"))]
#[test_case("strings_map", include_str!("../../../testdata/valid/strings_map.rhai"))]
#[test_case("switch", include_str!("../../../testdata/valid/switch.rhai"))]
#[test_case("while", include_str!("../../../testdata/valid/while.rhai"))]
#[test_case("char", include_str!("../../../testdata/valid/char.rhai"))]
#[test_case("throw_try_catch", include_str!("../../../testdata/valid/throw_try_catch.rhai"))]
#[test_case("optional_ops", include_str!("../../../testdata/valid/optional_ops.rhai"))]
fn parse_valid(name: &str, src: &str) {
    let parse = Parser::new(src).parse_script();
    assert!(parse.errors.is_empty(), "{:#?}", parse.errors);

    let mut engine = rhai::Engine::new();
    engine.set_max_expr_depths(0, 0);

    if src.starts_with("#!") {
        engine
            .compile(
                src.lines()
                    .skip(1)
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join("\n"),
            )
            .unwrap();
    } else {
        engine.compile(src).unwrap();
    }

    insta::with_settings!(
        { snapshot_suffix => name },
        {
            insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));
        }
    );
}

#[test]
fn parse_custom_operator() {
    let src = include_str!("../../../testdata/valid/operators.rhai");

    let parse = Parser::new(src)
        .with_operator("over", Operator::default())
        .parse_script();
    assert!(parse.errors.is_empty(), "{:#?}", parse.errors);

    insta::assert_snapshot!(format!("{:#?}", parse.into_syntax()));

    let parse = Parser::new(src).parse_script();
    assert!(!parse.errors.is_empty());
}
