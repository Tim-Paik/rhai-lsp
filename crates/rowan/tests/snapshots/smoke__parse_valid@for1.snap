---
source: crates/rowan/tests/smoke.rs
assertion_line: 65
expression: "format!(\"{:#?}\", parse.into_syntax())"
---
RHAI@0..523
  COMMENT_LINE@0..30 "// This script runs f ..."
  WHITESPACE@30..32 "\n\n"
  STMT@32..77
    ITEM@32..76
      EXPR@32..76
        EXPR_LET@32..76
          KW_LET@32..35 "let"
          WHITESPACE@35..36 " "
          IDENT@36..39 "arr"
          WHITESPACE@39..40 " "
          OP_ASSIGN@40..41 "="
          EXPR@41..76
            WHITESPACE@41..42 " "
            EXPR_ARRAY@42..76
              PUNCT_BRACKET_START@42..43 "["
              EXPR@43..44
                EXPR_LIT@43..44
                  LIT@43..44
                    LIT_INT@43..44 "1"
              PUNCT_COMMA@44..45 ","
              WHITESPACE@45..46 " "
              EXPR@46..50
                EXPR_LIT@46..50
                  LIT@46..50
                    LIT_BOOL@46..50 "true"
              PUNCT_COMMA@50..51 ","
              WHITESPACE@51..52 " "
              EXPR@52..59
                EXPR_LIT@52..59
                  LIT@52..59
                    LIT_FLOAT@52..59 "123.456"
              PUNCT_COMMA@59..60 ","
              WHITESPACE@60..61 " "
              EXPR@61..68
                EXPR_LIT@61..68
                  LIT@61..68
                    LIT_STR@61..68 "\"hello\""
              PUNCT_COMMA@68..69 ","
              WHITESPACE@69..70 " "
              EXPR@70..71
                EXPR_LIT@70..71
                  LIT@70..71
                    LIT_INT@70..71 "3"
              PUNCT_COMMA@71..72 ","
              WHITESPACE@72..73 " "
              EXPR@73..75
                EXPR_LIT@73..75
                  LIT@73..75
                    LIT_INT@73..75 "42"
              PUNCT_BRACKET_END@75..76 "]"
    PUNCT_SEMI@76..77 ";"
  WHITESPACE@77..79 "\n\n"
  COMMENT_LINE@79..110 "// Loop over array wi ..."
  WHITESPACE@110..111 "\n"
  STMT@111..454
    ITEM@111..292
      EXPR@111..292
        EXPR_FOR@111..292
          KW_FOR@111..114 "for"
          PAT@114..121
            WHITESPACE@114..115 " "
            PAT_TUPLE@115..121
              PUNCT_PAREN_START@115..116 "("
              PARAM@116..117
                IDENT@116..117 "a"
              PUNCT_COMMA@117..118 ","
              WHITESPACE@118..119 " "
              PARAM@119..120
                IDENT@119..120 "i"
              PUNCT_PAREN_END@120..121 ")"
          WHITESPACE@121..122 " "
          KW_IN@122..124 "in"
          EXPR@124..129
            WHITESPACE@124..125 " "
            EXPR_IDENT@125..129
              IDENT@125..128 "arr"
              WHITESPACE@128..129 " "
          EXPR_BLOCK@129..292
            PUNCT_BRACE_START@129..130 "{"
            WHITESPACE@130..135 "\n    "
            STMT@135..270
              ITEM@135..264
                EXPR@135..264
                  EXPR_FOR@135..264
                    KW_FOR@135..138 "for"
                    PAT@138..145
                      WHITESPACE@138..139 " "
                      PAT_TUPLE@139..145
                        PUNCT_PAREN_START@139..140 "("
                        PARAM@140..141
                          IDENT@140..141 "b"
                        PUNCT_COMMA@141..142 ","
                        WHITESPACE@142..143 " "
                        PARAM@143..144
                          IDENT@143..144 "j"
                        PUNCT_PAREN_END@144..145 ")"
                    WHITESPACE@145..146 " "
                    KW_IN@146..148 "in"
                    EXPR@148..177
                      WHITESPACE@148..149 " "
                      EXPR_ARRAY@149..176
                        PUNCT_BRACKET_START@149..150 "["
                        EXPR@150..153
                          EXPR_LIT@150..153
                            LIT@150..153
                              LIT_CHAR@150..153 "'x'"
                        PUNCT_COMMA@153..154 ","
                        WHITESPACE@154..155 " "
                        EXPR@155..157
                          EXPR_LIT@155..157
                            LIT@155..157
                              LIT_INT@155..157 "42"
                        PUNCT_COMMA@157..158 ","
                        WHITESPACE@158..159 " "
                        EXPR@159..161
                          EXPR_PAREN@159..161
                            PUNCT_PAREN_START@159..160 "("
                            PUNCT_PAREN_END@160..161 ")"
                        PUNCT_COMMA@161..162 ","
                        WHITESPACE@162..163 " "
                        EXPR@163..166
                          EXPR_LIT@163..166
                            LIT@163..166
                              LIT_INT@163..166 "123"
                        PUNCT_COMMA@166..167 ","
                        WHITESPACE@167..168 " "
                        EXPR@168..170
                          EXPR_LIT@168..170
                            LIT@168..170
                              LIT_INT@168..170 "99"
                        PUNCT_COMMA@170..171 ","
                        WHITESPACE@171..172 " "
                        EXPR@172..175
                          EXPR_LIT@172..175
                            LIT@172..175
                              LIT_FLOAT@172..175 "0.5"
                        PUNCT_BRACKET_END@175..176 "]"
                      WHITESPACE@176..177 " "
                    EXPR_BLOCK@177..264
                      PUNCT_BRACE_START@177..178 "{"
                      WHITESPACE@178..187 "\n        "
                      STMT@187..221
                        ITEM@187..221
                          EXPR@187..221
                            EXPR_IF@187..221
                              KW_IF@187..189 "if"
                              EXPR@189..198
                                EXPR_BINARY@189..198
                                  EXPR@189..192
                                    WHITESPACE@189..190 " "
                                    EXPR_IDENT@190..192
                                      IDENT@190..191 "b"
                                      WHITESPACE@191..192 " "
                                  OP_GT@192..193 ">"
                                  EXPR@193..198
                                    WHITESPACE@193..194 " "
                                    EXPR_LIT@194..197
                                      LIT@194..197
                                        LIT_INT@194..197 "100"
                                    WHITESPACE@197..198 " "
                              EXPR_BLOCK@198..211
                                PUNCT_BRACE_START@198..199 "{"
                                WHITESPACE@199..200 " "
                                STMT@200..209
                                  ITEM@200..208
                                    EXPR@200..208
                                      EXPR_CONTINUE@200..208
                                        KW_CONTINUE@200..208 "continue"
                                  PUNCT_SEMI@208..209 ";"
                                WHITESPACE@209..210 " "
                                PUNCT_BRACE_END@210..211 "}"
                              WHITESPACE@211..221 "\n\n        "
                      STMT@221..258
                        ITEM@221..257
                          EXPR@221..257
                            EXPR_CALL@221..257
                              EXPR@221..226
                                EXPR_IDENT@221..226
                                  IDENT@221..226 "print"
                              ARG_LIST@226..257
                                PUNCT_PAREN_START@226..227 "("
                                EXPR@227..256
                                  EXPR_LIT@227..256
                                    LIT@227..256
                                      LIT_STR_TEMPLATE@227..256
                                        LIT_STR@227..229 "`("
                                        INTERPOLATION_START@229..231 "${"
                                        LIT_STR_TEMPLATE_INTERPOLATION@231..232
                                          STMT@231..232
                                            ITEM@231..232
                                              EXPR@231..232
                                                EXPR_IDENT@231..232
                                                  IDENT@231..232 "i"
                                        PUNCT_BRACE_END@232..233 "}"
                                        LIT_STR@233..235 ", "
                                        INTERPOLATION_START@235..237 "${"
                                        LIT_STR_TEMPLATE_INTERPOLATION@237..238
                                          STMT@237..238
                                            ITEM@237..238
                                              EXPR@237..238
                                                EXPR_IDENT@237..238
                                                  IDENT@237..238 "j"
                                        PUNCT_BRACE_END@238..239 "}"
                                        LIT_STR@239..244 ") = ("
                                        INTERPOLATION_START@244..246 "${"
                                        LIT_STR_TEMPLATE_INTERPOLATION@246..247
                                          STMT@246..247
                                            ITEM@246..247
                                              EXPR@246..247
                                                EXPR_IDENT@246..247
                                                  IDENT@246..247 "a"
                                        PUNCT_BRACE_END@247..248 "}"
                                        LIT_STR@248..250 ", "
                                        INTERPOLATION_START@250..252 "${"
                                        LIT_STR_TEMPLATE_INTERPOLATION@252..253
                                          STMT@252..253
                                            ITEM@252..253
                                              EXPR@252..253
                                                EXPR_IDENT@252..253
                                                  IDENT@252..253 "b"
                                        PUNCT_BRACE_END@253..254 "}"
                                        LIT_STR@254..256 ")`"
                                PUNCT_PAREN_END@256..257 ")"
                        PUNCT_SEMI@257..258 ";"
                      WHITESPACE@258..263 "\n    "
                      PUNCT_BRACE_END@263..264 "}"
              WHITESPACE@264..270 "\n\n    "
            STMT@270..291
              ITEM@270..291
                EXPR@270..291
                  EXPR_IF@270..291
                    KW_IF@270..272 "if"
                    EXPR@272..280
                      EXPR_BINARY@272..280
                        EXPR@272..275
                          WHITESPACE@272..273 " "
                          EXPR_IDENT@273..275
                            IDENT@273..274 "a"
                            WHITESPACE@274..275 " "
                        OP_EQ@275..277 "=="
                        EXPR@277..280
                          WHITESPACE@277..278 " "
                          EXPR_LIT@278..279
                            LIT@278..279
                              LIT_INT@278..279 "3"
                          WHITESPACE@279..280 " "
                    EXPR_BLOCK@280..290
                      PUNCT_BRACE_START@280..281 "{"
                      WHITESPACE@281..282 " "
                      STMT@282..288
                        ITEM@282..287
                          EXPR@282..287
                            EXPR_BREAK@282..287
                              KW_BREAK@282..287 "break"
                        PUNCT_SEMI@287..288 ";"
                      WHITESPACE@288..289 " "
                      PUNCT_BRACE_END@289..290 "}"
                    WHITESPACE@290..291 "\n"
            PUNCT_BRACE_END@291..292 "}"
    WHITESPACE@292..293 "\n"
    COMMENT_LINE@293..386 "//print(a);           ..."
    WHITESPACE@386..415 "\n                     ..."
    COMMENT_LINE@415..452 "//    because 'a' is  ..."
    WHITESPACE@452..454 "\n\n"
  STMT@454..523
    ITEM@454..522
      EXPR@454..522
        EXPR_FOR@454..522
          KW_FOR@454..457 "for"
          PAT@457..459
            WHITESPACE@457..458 " "
            PAT_IDENT@458..459
              IDENT@458..459 "i"
          WHITESPACE@459..460 " "
          KW_IN@460..462 "in"
          EXPR@462..479
            EXPR_CALL@462..478
              EXPR@462..468
                WHITESPACE@462..463 " "
                EXPR_IDENT@463..468
                  IDENT@463..468 "range"
              ARG_LIST@468..478
                PUNCT_PAREN_START@468..469 "("
                EXPR@469..470
                  EXPR_LIT@469..470
                    LIT@469..470
                      LIT_INT@469..470 "5"
                PUNCT_COMMA@470..471 ","
                WHITESPACE@471..472 " "
                EXPR@472..473
                  EXPR_LIT@472..473
                    LIT@472..473
                      LIT_INT@472..473 "0"
                PUNCT_COMMA@473..474 ","
                WHITESPACE@474..475 " "
                EXPR@475..477
                  EXPR_UNARY@475..477
                    OP_SUB@475..476 "-"
                    EXPR@476..477
                      EXPR_LIT@476..477
                        LIT@476..477
                          LIT_INT@476..477 "1"
                PUNCT_PAREN_END@477..478 ")"
            WHITESPACE@478..479 " "
          EXPR_BLOCK@479..522
            PUNCT_BRACE_START@479..480 "{"
            WHITESPACE@480..482 "  "
            COMMENT_LINE@482..506 "// runs from 5 down to 1"
            WHITESPACE@506..511 "\n    "
            STMT@511..520
              ITEM@511..519
                EXPR@511..519
                  EXPR_CALL@511..519
                    EXPR@511..516
                      EXPR_IDENT@511..516
                        IDENT@511..516 "print"
                    ARG_LIST@516..519
                      PUNCT_PAREN_START@516..517 "("
                      EXPR@517..518
                        EXPR_IDENT@517..518
                          IDENT@517..518 "i"
                      PUNCT_PAREN_END@518..519 ")"
              PUNCT_SEMI@519..520 ";"
            WHITESPACE@520..521 "\n"
            PUNCT_BRACE_END@521..522 "}"
    WHITESPACE@522..523 "\n"

